pvlib/solarposition.py

```diff
@@ -1210,3 +1210,84 @@ def hour_angle(times, longitude, equation_of_time):
     )).total_seconds() / 3600. for t in times])
     timezone = times.tz.utcoffset(times).total_seconds() / 3600.
     return 15. * (hours - 12. - timezone) + longitude + equation_of_time / 4.
+
+
+def _hours(times, hour_angle, longitude, equation_of_time):
+    """helper that converts hour angles in degrees into hours"""
+    tz_info = times.tz  # pytz timezone info
+    tz = tz_info.utcoffset(times).total_seconds() / 3600.
+    return (hour_angle - longitude - equation_of_time / 4.) / 15. + 12. + tz
+
+
+def _times(times, hours):
+    """helper converts hours from an array of floats to localized times"""
+    tz_info = times.tz
+    try:
+        hours = hours.values  # make a new reference
+    except AttributeError:
+        pass
+    try:
+        # OLD pandas, "TypeError: Already tz-aware, use tz_convert to convert."
+        times.tz = None
+    except AttributeError:
```

Comment:
I also don't understand this try/except. Your API declaration requires a localized DatetimeIndex, so you can use `times.tz_convert` is guaranteed to work. This is true for pandas 0.14.

Reply:
`tz_convert` changes the timezone from one to another, ie: it applies the `tz_utcoffset` so eg: from `Etc/GMT+8` to `UTC` would add 8 hours to all of the time.

But that's not what we want. We just want to make the tz-aware time a local naive time, so we want to effectively replace `tz_info` with `None`.

Again in some older version of pandas, you could just do this by setting `tz=None`, and in newer ones you can call `tz_localize(None)` to convert a tz-aware time to a naive one, *without* applying a utc offset.

EG: `2017-01-01T12:30:00-0800` just becomes `2017-01-01T12:30:00` same time, but no more timezone

Reply:
I understand now, thanks. 

> newer ones you can call tz_localize(None) to convert a tz-aware time to a naive one, without applying a utc offset.

Is this a documented behavior of `tz_localize`? I can't find official support for it. If not I recommend a different implementation.

Reply:
@wholmgren it's in the pandas documentation for [`DatetimeIndex.tz_localize`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DatetimeIndex.tz_localize.html#pandas.DatetimeIndex.tz_localize)
>Passing `None` will remove the time zone information preserving local time.

It's and also [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.dt.tz_localize.html#pandas.Series.dt.tz_localize):
>Passing `None` will remove the time zone information preserving local time.

and [there](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Timestamp.tz_localize.html)
>None will remove timezone holding local time.

but neither [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.tz_localize.html#pandas.Series.tz_localize)

nor [there](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.tz_localize.html#pandas.DataFrame.tz_localize)

@cwhanse [`normalize`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.dt.normalize.html#pandas.Series.dt.normalize) seems very interesting, and perhaps what we're both looking for? [Since v0.16](https://github.com/pandas-dev/pandas/pull/10047), see [what's new](https://pandas.pydata.org/pandas-docs/version/0.16/whatsnew.html).
>Convert times to midnight.
>The time component of the date-timeise converted to midnight i.e. 00:00:00. This is useful in cases, when the time does not matter. Length is unaltered. The timezones are unaffected.